File: server/probe-telegram.js
// То же самое, но для контейнера основного сайта: там node, а не python.
const dns = require('node:dns');
const net = require('node:net');
const tls = require('node:tls');
const fs = require('node:fs');

const HOST = 'api.telegram.org';

try {
  console.log('резолвер:');
  for (const line of fs.readFileSync('/etc/resolv.conf', 'utf8').split('\n')) {
    if (line.trim() && !line.startsWith('#')) console.log('   ', line);
  }
} catch (e) {
  console.log('    не прочитался:', e.message);
}

const check = (address, family) =>
  new Promise((resolve) => {
    const started = Date.now();
    const socket = net.connect({ host: address, port: 443, family }, () => {
      const tcp = Date.now() - started;
      const secure = tls.connect({ socket, servername: HOST }, () => {
        console.log(`    ${address.padEnd(40)} TCP ${tcp} мс, TLS поднялся`);
        secure.destroy();
        resolve();
      });
      secure.on('error', (e) => {
        console.log(`    ${address.padEnd(40)} TLS не поднялся: ${e.message}`);
        resolve();
      });
    });
    socket.setTimeout(5000, () => {
      console.log(`    ${address.padEnd(40)} НЕТ: таймаут (${Date.now() - started} мс)`);
      socket.destroy();
      resolve();
    });
    socket.on('error', (e) => {
      console.log(`    ${address.padEnd(40)} НЕТ: ${e.message} (${Date.now() - started} мс)`);
      resolve();
    });
  });

(async () => {
  for (const [family, name] of [[4, 'IPv4'], [6, 'IPv6']]) {
    let addresses = [];
    try {
      addresses = (await dns.promises.resolve(HOST, family === 4 ? 'A' : 'AAAA')) ?? [];
    } catch (e) {
      console.log(`${name}: резолв не удался — ${e.message}`);
      continue;
    }
    console.log(`${name}: адреса — ${addresses.join(', ') || 'нет'}`);
    for (const address of addresses) await check(address, family);
  }
})();

File: server/probe-telegram.py
"""Проверка пути до api.telegram.org изнутри контейнера. Ничего не меняет.

Печатает: какие адреса вернул резолвер, и по каждому — доходит ли
TCP-соединение и поднимается ли TLS. Именно TLS: фильтрация умеет принять
TCP и оборвать соединение уже на рукопожатии, и один connect() дал бы
ложное «работает».
"""
import socket
import ssl
import time

HOST = "api.telegram.org"
FAMILIES = ((socket.AF_INET, "IPv4"), (socket.AF_INET6, "IPv6"))

print("резолвер:")
try:
    with open("/etc/resolv.conf", encoding="utf-8") as fh:
        for line in fh:
            if line.strip() and not line.startswith("#"):
                print("   ", line.rstrip())
except OSError as exc:
    print("    не прочитался:", exc)

ctx = ssl.create_default_context()
for family, name in FAMILIES:
    try:
        infos = socket.getaddrinfo(HOST, 443, family, socket.SOCK_STREAM)
        addresses = sorted({info[4][0] for info in infos})
    except OSError as exc:
        print("%s: резолв не удался — %s" % (name, exc))
        continue
    print("%s: адреса — %s" % (name, ", ".join(addresses) or "нет"))
    for address in addresses:
        started = time.monotonic()
        sock = None
        try:
            sock = socket.socket(family, socket.SOCK_STREAM)
            sock.settimeout(5)
            sock.connect((address, 443))
            tcp = int((time.monotonic() - started) * 1000)
            wrapped = ctx.wrap_socket(sock, server_hostname=HOST)
            wrapped.close()
            sock = None
            print("    %-40s TCP %d мс, TLS поднялся" % (address, tcp))
        except Exception as exc:  # noqa: BLE001
            print(
                "    %-40s НЕТ: %s (%d мс)"
                % (address, exc, int((time.monotonic() - started) * 1000))
            )
        finally:
            if sock is not None:
                try:
                    sock.close()
                except OSError:
                    pass
